File: bezier/src/org/brogan/bezier/CurveFitter.java
package org.brogan.bezier;

import java.awt.geom.Point2D;
import java.util.ArrayList;
import java.util.List;

/**
 * Schneider curve-fitting algorithm: fits a polyline of sampled points to a
 * sequence of cubic Bézier segments.
 *
 * Reference: Philip J. Schneider, "An Algorithm for Automatically Fitting
 * Digitized Curves", Graphics Gems I (1990).
 *
 * Public API:
 *   Point2D.Double[] fit(List<Point2D.Double> pts, double errorThreshold)
 *
 * Returns a flat array [a0,c1,c2,a1, a1,c1',c2',a1', ...] — N*4 points for
 * N segments. Adjacent segments share the same end/start anchor value.
 */
public class CurveFitter {

    // ── Public entry point ────────────────────────────────────────────────────

    /**
     * Fit the given polyline to one or more cubic Bézier segments.
     *
     * @param pts            digitised input points (raw mouse positions)
     * @param errorThreshold maximum allowed pixel distance from any sample to the
     *                       fitted curve. Values 1–50 work well; higher → fewer
     *                       segments (looser fit).
     * @return flat array of control points: groups of 4 (a0,c1,c2,a1) per segment,
     *         or null if the input is too short.
     */
    public static Point2D.Double[] fit(List<Point2D.Double> pts, double errorThreshold) {
        List<Point2D.Double> d = removeDuplicates(pts);
        if (d.size() < 2) return null;

        // Pre-simplify with Douglas-Peucker: remove near-collinear intermediate points
        // that lie within errorThreshold of the simplified polyline.  This prevents the
        // Schneider fitter from recursing down to degenerate 2-point base cases on
        // nearly-straight tails (e.g. the user slowing down near the end of a stroke).
        d = douglasPeucker(d, Math.max(errorThreshold, 2.0));
        if (d.size() < 2) return null;

        // error is stored as squared distance throughout
        double err2 = errorThreshold * errorThreshold;

        List<Point2D.Double[]> segments = new ArrayList<>();

        if (d.size() == 2) {
            // Single straight-line segment
            Point2D.Double p0 = d.get(0), p1 = d.get(1);
            Point2D.Double c1 = new Point2D.Double(
                p0.x + (p1.x - p0.x) / 3.0, p0.y + (p1.y - p0.y) / 3.0);
            Point2D.Double c2 = new Point2D.Double(
                p0.x + 2.0 * (p1.x - p0.x) / 3.0, p0.y + 2.0 * (p1.y - p0.y) / 3.0);
            segments.add(new Point2D.Double[]{p0, c1, c2, p1});
        } else {
            Point2D.Double tHat1 = computeLeftTangent(d, 0);
            Point2D.Double tHat2 = computeRightTangent(d, d.size() - 1);
            fitCubic(d, 0, d.size() - 1, tHat1, tHat2, err2, segments);
        }

        // Flatten to a single array
        Point2D.Double[] result = new Point2D.Double[segments.size() * 4];
        for (int i = 0; i < segments.size(); i++) {
            Point2D.Double[] seg = segments.get(i);
            result[i * 4]     = seg[0];
            result[i * 4 + 1] = seg[1];
            result[i * 4 + 2] = seg[2];
            result[i * 4 + 3] = seg[3];
        }
        return result;
    }

    // ── Core recursive fitter ─────────────────────────────────────────────────

    private static void fitCubic(List<Point2D.Double> d,
                                  int first, int last,
                                  Point2D.Double tHat1, Point2D.Double tHat2,
                                  double err2,
                                  List<Point2D.Double[]> result) {
        if (last - first == 1) {
            // Only two points — emit a straight segment
            Point2D.Double p0 = d.get(first), p1 = d.get(last);
            double dist = distance(p0, p1) / 3.0;
            Point2D.Double c1 = new Point2D.Double(
                p0.x + tHat1.x * dist, p0.y + tHat1.y * dist);
            Point2D.Double c2 = new Point2D.Double(
                p1.x + tHat2.x * dist, p1.y + tHat2.y * dist);
            result.add(new Point2D.Double[]{p0, c1, c2, p1});
            return;
        }

        double[] u      = chordLengthParam(d, first, last);
        Point2D.Double[] bezier = generateBezier(d, first, last, u, tHat1, tHat2);

        double[] me    = maxError(d, first, last, bezier, u);
        double maxDist = me[0];
        int splitIdx   = (int) me[1];

        // Accept only if within error AND not looping.
        // wouldLoop() detects when control arms together overshoot the chord,
        // which produces self-intersecting cubics on highly curved arcs.
        if (maxDist < err2 && !wouldLoop(bezier)) {
            result.add(bezier);
            return;
        }

        // Attempt Newton-Raphson reparameterisation only when the initial fit is
        // not looping (a looping bezier reparameterises to another looping bezier).
        if (!wouldLoop(bezier) && maxDist < err2 * 4.0) {
            double[] uPrime   = reparameterize(d, first, last, u, bezier);
            Point2D.Double[] bezier2 = generateBezier(d, first, last, uPrime, tHat1, tHat2);
            double[] me2      = maxError(d, first, last, bezier2, uPrime);
            if (me2[0] < err2 && !wouldLoop(bezier2)) {
                result.add(bezier2);
                return;
            }
            splitIdx = (int) me2[1];
        }

        // Split at the point of maximum error and recurse
        Point2D.Double tCenter    = centerTangent(d, splitIdx);
        Point2D.Double tCenterNeg = new Point2D.Double(-tCenter.x, -tCenter.y);
        fitCubic(d, first,    splitIdx, tHat1,       tCenter,    err2, result);
        fitCubic(d, splitIdx, last,     tCenterNeg,  tHat2,      err2, result);
    }

    // ── Chord-length parameterisation ─────────────────────────────────────────

    private static double[] chordLengthParam(List<Point2D.Double> d, int first, int last) {
        int count = last - first + 1;
        double[] u = new double[count];
        u[0] = 0.0;
        for (int i = 1; i < count; i++) {
            u[i] = u[i - 1] + distance(d.get(first + i - 1), d.get(first + i));
        }
        double total = u[count - 1];
        if (total > 0) {
            for (int i = 1; i < count; i++) u[i] /= total;
        }
        u[count - 1] = 1.0; // ensure last point is exactly 1
        return u;
    }

    // ── Least-squares Bézier generation ──────────────────────────────────────

    private static Point2D.Double[] generateBezier(List<Point2D.Double> d,
                                                    int first, int last,
                                                    double[] u,
                                                    Point2D.Double tHat1,
                                                    Point2D.Double tHat2) {
        Point2D.Double p0 = d.get(first);
        Point2D.Double p3 = d.get(last);
        int count = last - first + 1;

        // A1[i], A2[i] are 2-vectors (x,y components stored as double[2])
        double[][] A1x = new double[count][1]; // we'll inline x/y
        double[] a1x = new double[count], a1y = new double[count];
        double[] a2x = new double[count], a2y = new double[count];

        for (int i = 0; i < count; i++) {
            double t = u[i];
            double b1 = B1(t), b2 = B2(t);
            a1x[i] = b1 * tHat1.x;
            a1y[i] = b1 * tHat1.y;
            a2x[i] = b2 * tHat2.x;
            a2y[i] = b2 * tHat2.y;
        }

        // Build the 2×2 system C * [alpha1; alpha2] = X
        double c00 = 0, c01 = 0, c11 = 0;
        double x0  = 0, x1  = 0;

        for (int i = 0; i < count; i++) {
            double t  = u[i];
            double b0 = B0(t), b3 = B3(t);

            c00 += a1x[i] * a1x[i] + a1y[i] * a1y[i];
            c01 += a1x[i] * a2x[i] + a1y[i] * a2y[i];
            c11 += a2x[i] * a2x[i] + a2y[i] * a2y[i];

            Point2D.Double pt = d.get(first + i);
            double tx = pt.x - (b0 * p0.x + b3 * p3.x);
            double ty = pt.y - (b0 * p0.y + b3 * p3.y);

            x0 += a1x[i] * tx + a1y[i] * ty;
            x1 += a2x[i] * tx + a2y[i] * ty;
        }

        double c10  = c01; // symmetric
        double det  = c00 * c11 - c01 * c10;
        double alpha1, alpha2;

        if (Math.abs(det) > 1e-12) {
            alpha1 = (x0 * c11 - c01 * x1) / det;
            alpha2 = (c00 * x1 - x0 * c10) / det;
        } else {
            // Degenerate: fallback to chord/3 heuristic
            double dist = distance(p0, p3) / 3.0;
            alpha1 = dist;
            alpha2 = dist;
        }

        // Guard against negative alphas (would flip tangent direction)
        if (alpha1 < 1e-6 || alpha2 < 1e-6) {
            double dist = distance(p0, p3) / 3.0;
            alpha1 = dist;
            alpha2 = dist;
        }

        Point2D.Double c1 = new Point2D.Double(
            p0.x + alpha1 * tHat1.x, p0.y + alpha1 * tHat1.y);
        Point2D.Double c2 = new Point2D.Double(
            p3.x + alpha2 * tHat2.x, p3.y + alpha2 * tHat2.y);

        return new Point2D.Double[]{p0, c1, c2, p3};
    }

    // ── Max error ─────────────────────────────────────────────────────────────

    /** Returns {maxSquaredDist, splitIndex}. */
    private static double[] maxError(List<Point2D.Double> d,
                                      int first, int last,
                                      Point2D.Double[] bezier,
                                      double[] u) {
        double maxDist = 0;
        int splitIdx   = (first + last) / 2;
        for (int i = first + 1; i < last; i++) {
            Point2D.Double pt = bezierPoint(bezier, u[i - first]);
            double dx   = pt.x - d.get(i).x;
            double dy   = pt.y - d.get(i).y;
            double dist = dx * dx + dy * dy;
            if (dist > maxDist) {
                maxDist  = dist;
                splitIdx = i;
            }
        }
        return new double[]{maxDist, splitIdx};
    }

    // ── Newton-Raphson reparameterisation ─────────────────────────────────────

    private static double[] reparameterize(List<Point2D.Double> d,
                                            int first, int last,
                                            double[] u,
                                            Point2D.Double[] bezier) {
        int count = last - first + 1;
        double[] uPrime = new double[count];
        for (int i = 0; i < count; i++) {
            uPrime[i] = newtonRaphsonStep(bezier, d.get(first + i), u[i]);
        }
        return uPrime;
    }

    /** One Newton-Raphson step to find the closest t on the cubic for point p. */
    private static double newtonRaphsonStep(Point2D.Double[] bezier,
                                             Point2D.Double p, double u) {
        Point2D.Double q  = bezierPoint(bezier, u);
        Point2D.Double q1 = bezierTangent(bezier, u);
        Point2D.Double q2 = bezierSecondDeriv(bezier, u);

        double numer = (q.x - p.x) * q1.x + (q.y - p.y) * q1.y;
        double denom = q1.x * q1.x + q1.y * q1.y
                     + (q.x - p.x) * q2.x + (q.y - p.y) * q2.y;
        if (Math.abs(denom) < 1e-12) return u;
        double uNew = u - numer / denom;
        if (uNew < 0) uNew = 0;
        if (uNew > 1) uNew = 1;
        return uNew;
    }

    // ── Tangent helpers ───────────────────────────────────────────────────────

    private static Point2D.Double computeLeftTangent(List<Point2D.Double> d, int end) {
        return normalize(sub(d.get(end + 1), d.get(end)));
    }

    private static Point2D.Double computeRightTangent(List<Point2D.Double> d, int end) {
        return normalize(sub(d.get(end - 1), d.get(end)));
    }

    private static Point2D.Double centerTangent(List<Point2D.Double> d, int center) {
        // Must point BACKWARD (toward center-1) to match the "right tangent" convention:
        // tHat2 for the left sub-curve and negated to tHat1 for the right sub-curve,
        // so that c2/c1 are placed on the interior side of the split point.
        return normalize(sub(d.get(center - 1), d.get(center + 1)));
    }

    // ── Bézier evaluation ─────────────────────────────────────────────────────

    private static Point2D.Double bezierPoint(Point2D.Double[] b, double t) {
        double b0 = B0(t), b1 = B1(t), b2 = B2(t), b3 = B3(t);
        return new Point2D.Double(
            b0 * b[0].x + b1 * b[1].x + b2 * b[2].x + b3 * b[3].x,
            b0 * b[0].y + b1 * b[1].y + b2 * b[2].y + b3 * b[3].y);
    }

    /** First derivative Q'(t). */
    private static Point2D.Double bezierTangent(Point2D.Double[] b, double t) {
        double mt = 1 - t;
        double k0 = 3 * mt * mt;
        double k1 = 6 * mt * t;
        double k2 = 3 * t  * t;
        return new Point2D.Double(
            k0 * (b[1].x - b[0].x) + k1 * (b[2].x - b[1].x) + k2 * (b[3].x - b[2].x),
            k0 * (b[1].y - b[0].y) + k1 * (b[2].y - b[1].y) + k2 * (b[3].y - b[2].y));
    }

    /** Second derivative Q''(t). */
    private static Point2D.Double bezierSecondDeriv(Point2D.Double[] b, double t) {
        double mt = 1 - t;
        double k0 = 6 * mt;
        double k1 = 6 * t;
        return new Point2D.Double(
            k0 * (b[2].x - 2 * b[1].x + b[0].x) + k1 * (b[3].x - 2 * b[2].x + b[1].x),
            k0 * (b[2].y - 2 * b[1].y + b[0].y) + k1 * (b[3].y - 2 * b[2].y + b[1].y));
    }

    // ── Bernstein basis polynomials ───────────────────────────────────────────

    private static double B0(double t) { double mt = 1 - t; return mt * mt * mt; }
    private static double B1(double t) { double mt = 1 - t; return 3 * mt * mt * t; }
    private static double B2(double t) { double mt = 1 - t; return 3 * mt * t  * t; }
    private static double B3(double t) { return t * t * t; }

    // ── Utility ───────────────────────────────────────────────────────────────

    /**
     * Douglas-Peucker polyline simplification.
     * Removes intermediate points that deviate less than {@code epsilon} pixels
     * from the straight line between their neighbours, recursively.
     */
    private static List<Point2D.Double> douglasPeucker(List<Point2D.Double> pts, double epsilon) {
        if (pts.size() <= 2) return new ArrayList<>(pts);
        Point2D.Double start = pts.get(0);
        Point2D.Double end   = pts.get(pts.size() - 1);
        double maxDist = 0;
        int    maxIdx  = 1;
        for (int i = 1; i < pts.size() - 1; i++) {
            double d = pointToLineDistance(pts.get(i), start, end);
            if (d > maxDist) { maxDist = d; maxIdx = i; }
        }
        if (maxDist > epsilon) {
            List<Point2D.Double> left  = douglasPeucker(pts.subList(0, maxIdx + 1), epsilon);
            List<Point2D.Double> right = douglasPeucker(pts.subList(maxIdx, pts.size()), epsilon);
            List<Point2D.Double> result = new ArrayList<>(left);
            result.addAll(right.subList(1, right.size())); // avoid duplicating split point
            return result;
        } else {
            List<Point2D.Double> result = new ArrayList<>();
            result.add(start);
            result.add(end);
            return result;
        }
    }

    /** Perpendicular distance from point {@code p} to the infinite line through {@code a} and {@code b}. */
    private static double pointToLineDistance(Point2D.Double p,
                                              Point2D.Double a, Point2D.Double b) {
        double dx = b.x - a.x, dy = b.y - a.y;
        double len2 = dx * dx + dy * dy;
        if (len2 < 1e-12) return distance(p, a);
        double t = ((p.x - a.x) * dx + (p.y - a.y) * dy) / len2;
        double cx = a.x + t * dx, cy = a.y + t * dy;
        double ex = p.x - cx,     ey = p.y - cy;
        return Math.sqrt(ex * ex + ey * ey);
    }

    private static List<Point2D.Double> removeDuplicates(List<Point2D.Double> pts) {
        List<Point2D.Double> result = new ArrayList<>();
        for (Point2D.Double p : pts) {
            if (result.isEmpty() || distance(result.get(result.size() - 1), p) > 0.01)
                result.add(p);
        }
        return result;
    }

    private static double distance(Point2D.Double a, Point2D.Double b) {
        double dx = b.x - a.x, dy = b.y - a.y;
        return Math.sqrt(dx * dx + dy * dy);
    }

    /**
     * Returns true if the cubic bezier would self-intersect (loop).
     * When the sum of the two control-arm lengths exceeds the chord, the curve
     * bends back on itself for arcs > ~142°. We force a split in that case.
     */
    private static boolean wouldLoop(Point2D.Double[] bezier) {
        double arm1  = distance(bezier[0], bezier[1]);
        double arm2  = distance(bezier[3], bezier[2]);
        double chord = distance(bezier[0], bezier[3]);
        return chord > 1.0 && (arm1 + arm2) > chord;
    }

    private static Point2D.Double sub(Point2D.Double a, Point2D.Double b) {
        return new Point2D.Double(a.x - b.x, a.y - b.y);
    }

    private static Point2D.Double normalize(Point2D.Double v) {
        double len = Math.sqrt(v.x * v.x + v.y * v.y);
        if (len < 1e-10) return new Point2D.Double(1, 0);
        return new Point2D.Double(v.x / len, v.y / len);
    }
}

File: bezier_py/canvas/curve_fitter.py
"""
CurveFitter — Schneider algorithm for fitting freehand strokes to cubic Bézier segments.

Port of CurveFitter.java.

Public API:
    fit(pts: list[QPointF], error_threshold: float) -> list[QPointF] | None

Returns a flat list [a0,c1,c2,a1, a1,c1',c2',a1', ...] — N*4 points per segment.
Adjacent segments share the same end/start anchor value.
"""
from __future__ import annotations
import math
from PySide6.QtCore import QPointF


# ── Internal 2D vector helpers (simple tuples for speed) ─────────────────────

def _dist(a: tuple, b: tuple) -> float:
    dx, dy = b[0] - a[0], b[1] - a[1]
    return math.sqrt(dx * dx + dy * dy)


def _sub(a: tuple, b: tuple) -> tuple:
    return (a[0] - b[0], a[1] - b[1])


def _normalize(v: tuple) -> tuple:
    length = math.sqrt(v[0] * v[0] + v[1] * v[1])
    if length < 1e-10:
        return (1.0, 0.0)
    return (v[0] / length, v[1] / length)


# ── Bernstein basis polynomials ───────────────────────────────────────────────

def _b0(t: float) -> float:
    mt = 1.0 - t; return mt * mt * mt

def _b1(t: float) -> float:
    mt = 1.0 - t; return 3.0 * mt * mt * t

def _b2(t: float) -> float:
    mt = 1.0 - t; return 3.0 * mt * t * t

def _b3(t: float) -> float:
    return t * t * t


def _bezier_point(b: list[tuple], t: float) -> tuple:
    b0, b1, b2, b3 = _b0(t), _b1(t), _b2(t), _b3(t)
    return (
        b0 * b[0][0] + b1 * b[1][0] + b2 * b[2][0] + b3 * b[3][0],
        b0 * b[0][1] + b1 * b[1][1] + b2 * b[2][1] + b3 * b[3][1],
    )


def _bezier_tangent(b: list[tuple], t: float) -> tuple:
    mt = 1.0 - t
    k0, k1, k2 = 3.0 * mt * mt, 6.0 * mt * t, 3.0 * t * t
    return (
        k0 * (b[1][0] - b[0][0]) + k1 * (b[2][0] - b[1][0]) + k2 * (b[3][0] - b[2][0]),
        k0 * (b[1][1] - b[0][1]) + k1 * (b[2][1] - b[1][1]) + k2 * (b[3][1] - b[2][1]),
    )


def _bezier_second_deriv(b: list[tuple], t: float) -> tuple:
    mt = 1.0 - t
    k0, k1 = 6.0 * mt, 6.0 * t
    return (
        k0 * (b[2][0] - 2.0 * b[1][0] + b[0][0]) + k1 * (b[3][0] - 2.0 * b[2][0] + b[1][0]),
        k0 * (b[2][1] - 2.0 * b[1][1] + b[0][1]) + k1 * (b[3][1] - 2.0 * b[2][1] + b[1][1]),
    )


# ── Douglas-Peucker polyline simplification ───────────────────────────────────

def _point_to_line_dist(p: tuple, a: tuple, b: tuple) -> float:
    dx, dy = b[0] - a[0], b[1] - a[1]
    len2 = dx * dx + dy * dy
    if len2 < 1e-12:
        return _dist(p, a)
    t = ((p[0] - a[0]) * dx + (p[1] - a[1]) * dy) / len2
    cx, cy = a[0] + t * dx, a[1] + t * dy
    ex, ey = p[0] - cx, p[1] - cy
    return math.sqrt(ex * ex + ey * ey)


def _douglas_peucker(pts: list[tuple], epsilon: float) -> list[tuple]:
    if len(pts) <= 2:
        return list(pts)
    start, end = pts[0], pts[-1]
    max_dist, max_idx = 0.0, 1
    for i in range(1, len(pts) - 1):
        d = _point_to_line_dist(pts[i], start, end)
        if d > max_dist:
            max_dist, max_idx = d, i
    if max_dist > epsilon:
        left  = _douglas_peucker(pts[:max_idx + 1], epsilon)
        right = _douglas_peucker(pts[max_idx:], epsilon)
        return left + right[1:]
    else:
        return [start, end]


def _remove_duplicates(pts: list[tuple]) -> list[tuple]:
    result: list[tuple] = []
    for p in pts:
        if not result or _dist(result[-1], p) > 0.01:
            result.append(p)
    return result


# ── Chord-length parameterisation ─────────────────────────────────────────────

def _chord_length_param(d: list[tuple], first: int, last: int) -> list[float]:
    count = last - first + 1
    u = [0.0] * count
    for i in range(1, count):
        u[i] = u[i - 1] + _dist(d[first + i - 1], d[first + i])
    total = u[-1]
    if total > 0:
        u = [x / total for x in u]
    u[-1] = 1.0
    return u


# ── Least-squares Bézier generation ──────────────────────────────────────────

def _generate_bezier(d: list[tuple], first: int, last: int,
                     u: list[float], t_hat1: tuple, t_hat2: tuple) -> list[tuple]:
    p0, p3 = d[first], d[last]
    count = last - first + 1

    a1x = [0.0] * count
    a1y = [0.0] * count
    a2x = [0.0] * count
    a2y = [0.0] * count

    for i in range(count):
        t = u[i]
        b1, b2 = _b1(t), _b2(t)
        a1x[i] = b1 * t_hat1[0]; a1y[i] = b1 * t_hat1[1]
        a2x[i] = b2 * t_hat2[0]; a2y[i] = b2 * t_hat2[1]

    c00 = c01 = c11 = 0.0
    x0 = x1 = 0.0

    for i in range(count):
        t = u[i]
        b0, b3 = _b0(t), _b3(t)
        c00 += a1x[i] * a1x[i] + a1y[i] * a1y[i]
        c01 += a1x[i] * a2x[i] + a1y[i] * a2y[i]
        c11 += a2x[i] * a2x[i] + a2y[i] * a2y[i]
        pt = d[first + i]
        tx = pt[0] - (b0 * p0[0] + b3 * p3[0])
        ty = pt[1] - (b0 * p0[1] + b3 * p3[1])
        x0 += a1x[i] * tx + a1y[i] * ty
        x1 += a2x[i] * tx + a2y[i] * ty

    c10 = c01
    det = c00 * c11 - c01 * c10

    if abs(det) > 1e-12:
        alpha1 = (x0 * c11 - c01 * x1) / det
        alpha2 = (c00 * x1 - x0 * c10) / det
    else:
        dist = _dist(p0, p3) / 3.0
        alpha1 = alpha2 = dist

    if alpha1 < 1e-6 or alpha2 < 1e-6:
        dist = _dist(p0, p3) / 3.0
        alpha1 = alpha2 = dist

    c1 = (p0[0] + alpha1 * t_hat1[0], p0[1] + alpha1 * t_hat1[1])
    c2 = (p3[0] + alpha2 * t_hat2[0], p3[1] + alpha2 * t_hat2[1])
    return [p0, c1, c2, p3]


# ── Max error ─────────────────────────────────────────────────────────────────

def _max_error(d: list[tuple], first: int, last: int,
               bezier: list[tuple], u: list[float]) -> tuple[float, int]:
    max_dist = 0.0
    split_idx = (first + last) // 2
    for i in range(first + 1, last):
        pt = _bezier_point(bezier, u[i - first])
        dx, dy = pt[0] - d[i][0], pt[1] - d[i][1]
        dist = dx * dx + dy * dy
        if dist > max_dist:
            max_dist = dist
            split_idx = i
    return max_dist, split_idx


# ── Newton-Raphson reparameterisation ─────────────────────────────────────────

def _newton_raphson_step(bezier: list[tuple], p: tuple, u: float) -> float:
    q  = _bezier_point(bezier, u)
    q1 = _bezier_tangent(bezier, u)
    q2 = _bezier_second_deriv(bezier, u)
    numer = (q[0] - p[0]) * q1[0] + (q[1] - p[1]) * q1[1]
    denom = (q1[0] * q1[0] + q1[1] * q1[1]
             + (q[0] - p[0]) * q2[0] + (q[1] - p[1]) * q2[1])
    if abs(denom) < 1e-12:
        return u
    u_new = u - numer / denom
    return max(0.0, min(1.0, u_new))


def _reparameterize(d: list[tuple], first: int, last: int,
                    u: list[float], bezier: list[tuple]) -> list[float]:
    count = last - first + 1
    return [_newton_raphson_step(bezier, d[first + i], u[i]) for i in range(count)]


# ── Loop detection ─────────────────────────────────────────────────────────────

def _would_loop(bezier: list[tuple]) -> bool:
    arm1  = _dist(bezier[0], bezier[1])
    arm2  = _dist(bezier[3], bezier[2])
    chord = _dist(bezier[0], bezier[3])
    return chord > 1.0 and (arm1 + arm2) > chord


# ── Tangent helpers ───────────────────────────────────────────────────────────

def _left_tangent(d: list[tuple], end: int) -> tuple:
    return _normalize(_sub(d[end + 1], d[end]))


def _right_tangent(d: list[tuple], end: int) -> tuple:
    return _normalize(_sub(d[end - 1], d[end]))


def _center_tangent(d: list[tuple], center: int) -> tuple:
    return _normalize(_sub(d[center - 1], d[center + 1]))


# ── Core recursive fitter ─────────────────────────────────────────────────────

def _fit_cubic(d: list[tuple], first: int, last: int,
               t_hat1: tuple, t_hat2: tuple,
               err2: float, result: list) -> None:
    if last - first == 1:
        p0, p1 = d[first], d[last]
        dist = _dist(p0, p1) / 3.0
        c1 = (p0[0] + t_hat1[0] * dist, p0[1] + t_hat1[1] * dist)
        c2 = (p1[0] + t_hat2[0] * dist, p1[1] + t_hat2[1] * dist)
        result.append([p0, c1, c2, p1])
        return

    u = _chord_length_param(d, first, last)
    bezier = _generate_bezier(d, first, last, u, t_hat1, t_hat2)

    max_dist, split_idx = _max_error(d, first, last, bezier, u)

    if max_dist < err2 and not _would_loop(bezier):
        result.append(bezier)
        return

    if not _would_loop(bezier) and max_dist < err2 * 4.0:
        u_prime = _reparameterize(d, first, last, u, bezier)
        bezier2 = _generate_bezier(d, first, last, u_prime, t_hat1, t_hat2)
        max_dist2, split_idx2 = _max_error(d, first, last, bezier2, u_prime)
        if max_dist2 < err2 and not _would_loop(bezier2):
            result.append(bezier2)
            return
        split_idx = split_idx2

    t_center = _center_tangent(d, split_idx)
    t_center_neg = (-t_center[0], -t_center[1])
    _fit_cubic(d, first, split_idx, t_hat1,       t_center,     err2, result)
    _fit_cubic(d, split_idx, last,  t_center_neg, t_hat2,       err2, result)


# ── Public API ────────────────────────────────────────────────────────────────

def fit(pts: list[QPointF], error_threshold: float) -> list[QPointF] | None:
    """
    Fit a freehand stroke to cubic Bézier segments.

    Parameters
    ----------
    pts             : raw sample points from mouse drag
    error_threshold : max pixel error per sample (1–50 typical; higher = fewer segments)

    Returns
    -------
    Flat list of QPointF in groups of 4 (a0, c1, c2, a1) per segment, or None if
    the input is too short.
    """
    d = _remove_duplicates([(p.x(), p.y()) for p in pts])
    if len(d) < 2:
        return None

    d = _douglas_peucker(d, max(error_threshold, 2.0))
    if len(d) < 2:
        return None

    err2 = error_threshold * error_threshold
    segments: list = []

    if len(d) == 2:
        p0, p1 = d[0], d[1]
        c1 = (p0[0] + (p1[0] - p0[0]) / 3.0, p0[1] + (p1[1] - p0[1]) / 3.0)
        c2 = (p0[0] + 2.0 * (p1[0] - p0[0]) / 3.0, p0[1] + 2.0 * (p1[1] - p0[1]) / 3.0)
        segments.append([p0, c1, c2, p1])
    else:
        t_hat1 = _left_tangent(d, 0)
        t_hat2 = _right_tangent(d, len(d) - 1)
        _fit_cubic(d, 0, len(d) - 1, t_hat1, t_hat2, err2, segments)

    result: list[QPointF] = []
    for seg in segments:
        for pt in seg:
            result.append(QPointF(pt[0], pt[1]))
    return result
